lib/putS3.ts

```diff
@@ -0,0 +1,101 @@
+type Warning = string;
+
+export async function putFiles(
+    project: Project,
+    inv: GoalInvocation,
+    s3: S3,
+    params: PublishToS3Options): Promise<[FilesAttempted, SuccessfullyPushedKey[], Warning[]]> {
+    const { bucketName, filesToPublish, pathTranslation } = params;
+    let fileCount = 0;
+    const log = inv.progressLog;
+    const keys: SuccessfullyPushedKey[] = [];
+    const warnings: Warning[] = [];
+    await doWithFiles(project, filesToPublish, async file => {
+        const key = pathTranslation(file.path, inv);
+        const contentType = mime.lookup(file.path) || "text/plain";
+        const content = await file.getContentBuffer();
+        let fileParams: Partial<S3.Types.PutObjectRequest> = {};
+        if (params.paramsExt) {
+            const [retrievedFileParams, additionalWarnings] = await gatherParamsFromCompanionFile(project, log, file, params.paramsExt);
+            fileParams = retrievedFileParams;
+            additionalWarnings.forEach(w => warnings.push(w));
+        }
+        const objectParams = {
+            Bucket: bucketName,
+            Key: key,
+            Body: content,
+            ContentType: contentType,
+            ...fileParams,
+        };
+        logger.debug(`File: ${file.path}, key: ${key}, contentType: ${contentType}`);
+        try {
+            await s3.putObject(objectParams).promise();
+            keys.push(key);
+            fileCount++;
+            log.write(`Put '${file.path}' to 's3://${bucketName}/${key}'`);
+        } catch (e) {
+            const msg = `Failed to put '${file.path}' to 's3://${bucketName}/${key}': ${e.code}: ${e.message}`;
+            log.write(msg);
+            warnings.push(msg);
+        }
+    });
+    return [fileCount, keys, warnings];
+}
+
+async function gatherParamsFromCompanionFile(project: Project,
+                                             log: ProgressLog,
+                                             file: ProjectFile,
+                                             companionFileExtension: string): Promise<[Partial<S3.Types.PutObjectRequest>, string[]]> {
+    const companionFilePrefix = ".";
+    const paramsPath = file.path.replace(escapeSpecialCharacters(file.name),
+        `${companionFilePrefix}${file.name}${companionFileExtension}`);
+    const paramsFile = await project.getFile(paramsPath);
+    if (!paramsFile) {
+        return [{}, []];
+    }
+    try {
+        const fileParams = JSON.parse(await paramsFile.getContent());
+        log.write(`Merging in S3 parameters from '${paramsPath}': ${JSON.stringify(fileParams)}`);
+        return [fileParams, []];
+    } catch (e) {
+        const msg = `Failed to read and parse S3 params file '${paramsPath}', using defaults: ${e.message}`;
+        log.write(msg);
+        return [{}, [msg]];
+    }
+}
+
+export function escapeSpecialCharacters(filename: string): RegExp {
+    return new RegExp(filename.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&") + "$");
```

Comment:
It is good to clarify this by extracting it into a descriptively-named function. Technically, the escaping of special characters is just the `filename.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&")` part. The rest is turning it into an anchored regular expression. I wonder if we just shouldn't replace this mess with a `slice` using the length of the file name.

Reply:
good idea